src/classes/AuthenticationSession.ts

```diff
@@ -13,25 +13,94 @@ enum FusionAuthStatusCode {
 }
 
 export class AuthenticationSession {
-  public authToken = '';
+  private authToken = '';
+
+  public refreshToken = '';
 
   public readonly authContext;
 
+  private authTokenExpiresAt = new Date();
+
   private readonly fusionAuthClient;
 
   private twoFactorId = '';
 
   private twoFactorMethods: TwoFactorMethod[] = [];
 
-  public constructor(authContext: KeyboardAuthContext) {
+  private fusionAuthSftpAppId = '';
+
+  private fusionAuthSftpClientId = '';
+
+  private fusionAuthSftpClientSecret = '';
+
+  public constructor(
+    authContext: KeyboardAuthContext,
+    fusionAuthSftpAppId: string,
+    fusionAuthSftpClientId: string,
+    fusionAuthSftpClientSecret: string,
+  ) {
     this.authContext = authContext;
+    this.fusionAuthSftpAppId = fusionAuthSftpAppId;
+    this.fusionAuthSftpClientId = fusionAuthSftpClientId;
+    this.fusionAuthSftpClientSecret = fusionAuthSftpClientSecret;
     this.fusionAuthClient = getFusionAuthClient();
   }
 
   public invokeAuthenticationFlow(): void {
     this.promptForPassword();
   }
 
+  public async getToken() {
+    if (this.tokenWouldExpireSoon()) {
+      await this.getAuthTokenUsingRefreshToken();
+    }
+    return this.authToken;
+  }
+
+  private async getAuthTokenUsingRefreshToken(): Promise<void> {
+    try {
+      const clientResponse = await this.fusionAuthClient.exchangeRefreshTokenForAccessToken(
+        this.refreshToken,
+        this.fusionAuthSftpClientId,
+        this.fusionAuthSftpClientSecret,
+        '',
+        '',
+      );
+
+      if (clientResponse.response.access_token) {
+        this.authToken = clientResponse.response.access_token;
+        // The exchange refresh token for access token endpoint does not return a timestamp,
+        // it returns expires_in in seconds.
+        // So we need to create the timestamp to be consistent with what is first
+        // returned upon initial authentication
+        this.authTokenExpiresAt = new Date(
+          Date.now() + (clientResponse.response.expires_in ?? 1 * 1000),
+        );
+        logger.info('New access token obtained :', clientResponse.response);
+      } else {
+        logger.warn('No refresh token in response :', clientResponse.response);
+        this.authContext.reject();
+      }
+    } catch (error: unknown) {
```

Comment:
Umm, the block is 28 lines. 7 lines are comments, 7 lines are the actual call, 12 lines are things that directly depends on the call succeeding and hence bundled in the try catch. 

Alternative is defined a null response out the try-catch and only catch the and then handle process checks after the catch which makes it less try-catch counter intuitive IMO.

Not marking as resolve, if there's more context to be thrown in. Thanks @slifty

Reply:
Just to be clear your try / catch block would still stop processing in the event of a failure -- so the code that relies on success does NOT need to be wrapped in the entire try block.  Right now the try block throws several errors which immediately get caught and re-thrown in the catch block.

Here's what I mean as an alternative (typed in github so may well have issues):

```
let clientResponse;
try {
      /**
       * Fusion auth sdk wrongly mandates last two params (scope, user_code)
       * hence the need to pass two empty strings here.
       * See: https://github.com/FusionAuth/fusionauth-typescript-client/issues/42
       */
      const clientResponse = await this.fusionAuthClient.exchangeRefreshTokenForAccessToken(
        this.refreshToken,
        this.fusionAuthClientId,
        this.fusionAuthClientSecret,
        '',
        '',
      );
} catch (error: unknown) {
      if (isPartialClientResponse(error)) {
        throw new AuthTokenRefreshError(`Error returned by FusionAuth while obtaining refresh token: ${message}`);
      }
      if (error instanceof Error) {
        throw new AuthTokenRefreshError(`Error generated when calling by FusionAuth to obtaining refresh token: ${error.message}`);
      }
      console.log(error);
      throw new AuthTokenRefreshError('Unexpected throw when calling FusionAuth to obtain refresh token');
}

if (clientResponse.response.access_token) {
  this.authToken = clientResponse.response.access_token;
  if (clientResponse.response.expires_in) {
    this.authTokenExpiresAt = new Date(
      Date.now() + (clientResponse.response.expires_in * 1000),
    );
    logger.info('New access token obtained :', clientResponse.response);
  } else {
    logger.error('Response lacking token TTL (expires_in)');
    throw new AuthTokenRefreshError('Response lacking token TTL (expires_in)');
  }
} else {
  logger.error('No refresh token in response :', clientResponse.response);
  throw new AuthTokenRefreshError('No refresh token in response');
}
```

By approaching it this way you can make sure the resulting thrown errors are very specific to exactly what went wrong; the larger try block means your catch block necessarily has to become more generic because it has to cover a wide variety of possible thrown errors (some of which are better passed directly with nothing in the middle)